### PR TITLE
DOCS add secure application access draft

### PR DESCRIPTION
**Pull request by fabianmuehlberger**
I started writing a tutorial regarding security of TrueNAS applications. 

I would have created an Issue for my proposal but created a fork and PR instead. Furthermore, I am not sure if this belongs in master since it is still a draft, but wanted some feedback. Since I am relatively new to TrueNAS it is likely that my contribution contains errors.

I for now I created a _index.md with the content, under Apps/AppSecurity and stored the images under src/static/SCALE/Apps. In frontmatter I added the author (me) and set draft: true

Guidance for further work is appreciated!

Thanks for contributing to TrueNAS documentation! By opening a Pull Request, you're acknowledging that your changes will be distributed under the [Creative Commons 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

### CODE REVIEW

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+## Cloudflare
+
+For this solution to work, a free Cloudflare account is needed, as well as registered domain. The Cloudflare documentation is very well written and covers everything you need to know, setting up a domain and [DNS](https://developers.cloudflare.com/dns/). 
+
+### Creating a Cloudflare Tunnel
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}
+
+The Cloudflare Tunnel is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:44]:
```suggestion
[Cloudflare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}

```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+### Creating a Cloudflare Tunnel
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}
+
+The Cloudflare Tunnel is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application. 
+
+It is highly recommended to read the [tunnel documentation](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) or to watch [this](https://www.youtube.com/watch?v=eojWaJQvqiw) from Lawrence Systems setting up Cloudflare tunnels for applications. This guide assumes that the applications run as a docker container but the same approach can be used to secure apps running on Truenas Scale in Kubernetes.
+
+In the Cloudflare One dashboard, set a public hostname for accessing Nextcloud, this hostname, in this example it is:  **nextcloud.example.com**
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:48]:
```suggestion
In the Cloudflare One dashboard, set a public hostname for accessing Nextcloud, for example:  *nextcloud.example.com*.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md (hunk 1 above)
Comment (DjP-iX) [reply content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:44]: Note we should add a credit to Cloudflare for the overview image

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+For this solution to work, a free Cloudflare account is needed, as well as registered domain. The Cloudflare documentation is very well written and covers everything you need to know, setting up a domain and [DNS](https://developers.cloudflare.com/dns/). 
+
+### Creating a Cloudflare Tunnel
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}
+
+The Cloudflare Tunnel is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application. 
+
+It is highly recommended to read the [tunnel documentation](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) or to watch [this](https://www.youtube.com/watch?v=eojWaJQvqiw) from Lawrence Systems setting up Cloudflare tunnels for applications. This guide assumes that the applications run as a docker container but the same approach can be used to secure apps running on Truenas Scale in Kubernetes.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:46]:
```suggestion
This guide assumes that the applications run as a docker container but the same approach can be used to secure apps running on Truenas Scale in Kubernetes.
 [This video](https://www.youtube.com/watch?v=eojWaJQvqiw) from Lawrence Systems provides a detailed overview of setting up Cloudflare tunnels for applications. 
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+In the Cloudflare One dashboard, set a public hostname for accessing Nextcloud, this hostname, in this example it is:  **nextcloud.example.com**
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelHostname.png" alt="Cloudflare Tunnel Hostname" id="Cloudflare Tunnel Hostname" >}}
+
+The service Type is set to **HTTPS** and the URL is the local TrueNAS IP + the Nextcloud container port, e.g. 192.168.1.1:9001
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:52]:
```suggestion
Set service **Type** to **HTTPS**.
Enter the local TrueNAS IP with the Nextcloud container port, for example *192.168.1.1:9001*.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelHostname.png" alt="Cloudflare Tunnel Hostname" id="Cloudflare Tunnel Hostname" >}}

```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelTLSSetting.png" alt="Cloudflare Tunnel TLS Setting" id="Cloudflare Tunnel TLS Setting" >}}
+
+Additional application Settings > TLS > No TLS Verify = Enabled
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:56]:
```suggestion
Go to **Additional application Settings**, select **TLS** from the dropdown menu, and enable **No TLS Verify**.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelTLSSetting.png" alt="Cloudflare Tunnel TLS Setting" id="Cloudflare Tunnel TLS Setting" >}}

```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+### Nextcloud Configuration
+
+Two environment variables must be set in the Nextcloud application.
+[overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost)
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:61]:
```suggestion
Install the [Nextcloud]({{< relref "/scale/scaletutorials/apps/communityapps/installnextcloudmedia.md" >}}) community application.

Click **Add** next to **Nextcloud environment** to configure the environment variables [overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost).
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+In the Nextcloud App the environment variables are edited as shown below, where `value` is the address.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:63]:
```suggestion
Enter the two environment variables in **Name** as *OVERWRITECLIURL* and *OVERWRITEHOST*.

Enter the address for the Cloudflare Tunnel configured above in **Value**, for example *nextcloud.example.com*.

```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+Additional application Settings > TLS > No TLS Verify = Enabled
+
+### Nextcloud Configuration
+
+Two environment variables must be set in the Nextcloud application.
+[overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost)
+
+In the Nextcloud App the environment variables are edited as shown below, where `value` is the address. 
+
+{{< trueimage src="/images/SCALE/Apps/NextcloudEnviromentVariables.png" alt="Nextcloud Environment Variables" id="Nextcloud Environment Variables" >}}
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:65]: At the end can we wrap up the procedure with a step for confirming the configuration succeeded? Maybe something like wait for the installed application to deploy and then access via the proxy URL?

**Review by DjP-iX — CHANGES_REQUESTED** [review]
Thank you for the excellent contribution @fabianmuehlberger!

I've left a number of (largely stylistic) suggestions, but the overall procedure looks good. 

Note that we're probably going to move this from an index to something like /apps/SecuringApps.md before publishing, but leaving it here for now will let you see the suggestions easier.

**Comment by fabianmuehlberger** [comment]
Hey @DjP-iX thank you for the prompt reply. Since I went through the process of setting up those applications (first time TrueNAS owner), I thought I write it down while I have it in memory. 

Preferably, if you like, I would commit to e.g. "DEV DOC appsecurity" branch created by you. Otherwise, I can stay on my fork if you prefer that. 

TY for the suggestions, I hope to implement them tomorrow. 

> Note that we're probably going to move this from an index to something like /apps/SecuringApps.md before publishing, but leaving it here for now will let you see the suggestions easier.

Regarding the doc structure, I intend to cover multiple ways of securing applications, including Cloudflare, Tailscale, etc. ideally in a more generalized approach, and with the help of the community.  
I am not sure about the document structure, depending on the covered areas and apps, I think it should be split into different pages, for easier editing. 

Best Fabian

**Comment by fabianmuehlberger** [comment]
Regarding the image credit, I tried to use "caption" 

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" **caption="cloudflare.com"** >}} but that does not work.  How should I give credit correctly?

**Comment by fabianmuehlberger** [comment]
I am not sure how to add a relative ref to an anchor.

line 24: See [security]{{< relref "_index.md#additional-security-considerations" >}}

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+### Testing the Cloudflare Tunnel
+
+With the Cloudflare connector and Nextcloud installed and configured, in your Cloudflare dashboard, go to **Networks** and select **Tunnes**.
+
+The status of the tunnel should be **HEALTY**
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelStateHealthy.png" alt="Cloudflare Tunnel Healty" id="Cloudflare Tunnel Healty" >}}
+
+Nextcloud should now be reachable via nextcloud.example.com and use a https connection.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:92]:
```suggestion
Nextcloud should now be reachable via nextcloud.example.com, using a https connection.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+### Additional Security Considerations 
+
+Use strong user passwords and cofigure 2FA for additional security. [Two Factor Authentication](https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html)
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:96]:
```suggestion
Use strong user passwords and configure [Two Factor Authentication](https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html) for additional security.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+Cloudflare offers Access policies to restrict access to the application to specific users, emails or authentication methods. You can find more information in the [applications documentation](https://developers.cloudflare.com/cloudflare-one/applications/).
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:98]:
```suggestion
Cloudflare offers [access policies](https://developers.cloudflare.com/cloudflare-one/policies/access/) to restrict access to the application to specific users, emails or authentication methods.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+Go to **Access**, click **Add an Application** button and select **Self-Hosted**
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:100]:
```suggestion
Go to **Access**, click **Add an Application**, and select **Self-Hosted**.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+### Additional Security Considerations 
+
+Use strong user passwords and cofigure 2FA for additional security. [Two Factor Authentication](https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html)
+
+Cloudflare offers Access policies to restrict access to the application to specific users, emails or authentication methods. You can find more information in the [applications documentation](https://developers.cloudflare.com/cloudflare-one/applications/).
+
+Go to **Access**, click **Add an Application** button and select **Self-Hosted** 
+
+Add your nextcloud application and the domain you configured in the Cloudflare tunnel.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:102]:
```suggestion
Add your Nextcloud application and the domain you configured in the Cloudflare tunnel.
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareAddApplication.png" alt="Cloudflare Add Application" id="Cloudflare Add Application" >}}
+
+Click **Next**. 
+
+Create a new new Policy by entering a **Policy Name**. Groups can be assigned to this policy or additional rules can be added.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:108]:
```suggestion
Create a new policy by entering a **Policy Name**. Groups can be assigned to this policy or additional rules can be added. 
```

File: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+Create a new new Policy by entering a **Policy Name**. Groups can be assigned to this policy or additional rules can be added. 
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareAddPolicy.png" alt="Cloudflare Add Policy" id="Cloudflare Add Policy" >}}
+
+{{< trueimage src="/images/SCALE/Apps/CloudflareCreateAdditionalRules.png" alt="Cloudflare Additional Rules" id="Cloudflare Additional Rules" >}}
+
+Click **Next** and **Save**.
+
+> Note, that there are additional options for configurations, that are beyond the scope of this tutorial.
```
Comment (DjP-iX) [inline content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md:116]:
```suggestion
Note: there are additional options for policy configuration, but these are beyond the scope of this tutorial.
```

**Review by DjP-iX — DISMISSED** [review]
Looking good @fabianmuehlberger!

If you're planning to contribute additional articles on this subject, then I think we are fine to leave this as an index instead of an individual article.

**Review by fabianmuehlberger — COMMENTED** [review]
For splitting the document, I suggest, having everything related to a specific app (in that case, Cloudflare separated).

I would keep it as it is, until there is new content.

**Review by DjP-iX — APPROVED** [review]

**Comment by DjP-iX** [comment]
Thanks again @fabianmuehlberger!

This one is merged, but feel free to continue working in your branch and open a PR against our Master branch if you have any further changes. Likewise if you do write any further content for other apps/solutions. We'll be happy to see your contributions.

**Comment by fabianmuehlberger** [comment]
Thank you for the support and lending me a hand! This was my first bigger contribution to a project.

**Pull request merged**